### ISSUE TITLE
Contexts proposal

### ISSUE THREAD

**Issue by AlexMost** [c-3po-org/c-3po#65]
One of the main feature request from other people, who sees `c-3po` is the `context` feature.
Context allows us to translate the same word (the same msgid) differently, depending on the context.

For C, C++ there are separate functions for applying context to the translation: pgettext, npgettext (for single and plural forms accordingly)

```
pgettext(<context>, <msgid>);
pngettext(<context>, <single form>, <plural form>, <count>);
```

As for me I don't like the idea of having separate functions for that. As it was designed for C initially, they almost had no other option. But we can make it better, so I would like to implement succinct API for the `c-3po`.

### Proposal 1
To introduce special `c` function, that will return an object with all c-3po functions with applied context. It will accept context value as a first argument.

```js
import { c, t } from 'c-3po'
c('context').t`Hello ${ user.name }`

// maybe better to place it on a separate lines
c('context')
.t`Hello ${ user.name}`
```

```js
import { ngettext, msgid, c } from 'c-3po'
c('context').ngettext(msgid`${ n } banana`, `${ n } bananas`, n);
```
One thing that I am worried about, is that syntax becomes a little bit overloaded. But in general, I think context is not so frequent case. And with this apporach, it is clear how to introduce some other concepts like `domains` by just chaining them one after another. 
```js
d('domain')
.c('context')
.t`translate me`
```
### Proposal 2
Place context information right into the `t` function call;
```js
import { t } from 'c-3po'
t('context')`Hello ${ user.name}`
```
```js
import { ngettext, msgid } from 'c-3po'
ngettext('context')(msgid`${ n } banana`, `${ n } bananas`, n);
```
This one looks a little bit cleaner that Proposal 1, but the question is how do we extend this to have domain info? Maybe something like this will work:
```js
t({context: 'context', domain: 'domain'})`Hello ${ user.name}`
```

What do you think @MrOrz @alxpy @Suhoy95

**Comment by AlexMost** [comment]
As for me, I like the first proposal more.

**Comment by Suhoy95** [comment]
Hm.. Yes, the first proposal looks better. 

I think 2 proposal is more unclear (or sophisticated). Gettext functions need to be as dumb as possible.

**Comment by MrOrz** [comment]
Both proposals looks good to me.
Since the first proposal does not modify the signature of `t()`, I think would be easier to maintain & implement.

**Comment by AlexMost** [comment]
@MrOrz thanks for the feedback, hope to have a chance to get started on this soon. BTW, have you encountered some real use case for the contexts? We have quite a big codebase on our project and still, there were no use cases for this feature. But I agree, that we definitely should support that.

**Comment by MrOrz** [comment]
I encountered this just today.

In our application, there is a form control that is similar to:

```
You may also include XXXX when doing YYY.

( Radio button ) Include XXX
( Radio button ) Do not include XXX
```

And in another page, we have a more concise version of the control, which contains just the options.

```
( Radio button ) Include XXX
( Radio button ) Do not include XXX
```

While it makes perfect sense in English, our Japanese translator suggests that the options in the first scenario would look nicer if translated without "XXX", while leaving the second scenario intact.

**Comment by AlexMost** [comment]
ready in `0.7.0-release` branch. Will be available soon in `0.7.0-beta`, stay tuned :D

**Comment by AlexMost** [comment]
available in `0.7.0` version.
doc - https://c-3po.js.org/contexts.html
